File: src/azure-communication-services-integrate-voice-live-api/dotnet/AudioResampler.cs
using System;

namespace Azure.Communication.CallAutomation
{
    public static class AudioResampler
    {
        /// <summary>
        /// Resamples PCM audio from 16KHz to 24KHz using linear interpolation
        /// </summary>
        /// <param name="input16kHz">Input PCM audio data at 16KHz (16-bit samples)</param>
        /// <returns>Resampled PCM audio data at 24KHz (16-bit samples)</returns>
        public static byte[] Resample16kTo24k(byte[] input16kHz)
        {
            if (input16kHz == null || input16kHz.Length == 0)
                return input16kHz;

            // Convert byte array to 16-bit samples
            int inputSampleCount = input16kHz.Length / 2;
            short[] inputSamples = new short[inputSampleCount];
            Buffer.BlockCopy(input16kHz, 0, inputSamples, 0, input16kHz.Length);

            // Calculate output sample count (24/16 = 1.5x more samples)
            int outputSampleCount = (int)(inputSampleCount * 1.5);
            short[] outputSamples = new short[outputSampleCount];

            // Resample using linear interpolation
            double ratio = 16000.0 / 24000.0; // 0.6667

            for (int i = 0; i < outputSampleCount; i++)
            {
                double srcIndex = i * ratio;
                int srcIndexInt = (int)srcIndex;
                double fraction = srcIndex - srcIndexInt;

                if (srcIndexInt >= inputSampleCount - 1)
                {
                    outputSamples[i] = inputSamples[inputSampleCount - 1];
                }
                else
                {
                    // Linear interpolation between two samples
                    short sample1 = inputSamples[srcIndexInt];
                    short sample2 = inputSamples[srcIndexInt + 1];
                    outputSamples[i] = (short)(sample1 + (sample2 - sample1) * fraction);
                }
            }

            // Convert back to byte array
            byte[] output24kHz = new byte[outputSamples.Length * 2];
            Buffer.BlockCopy(outputSamples, 0, output24kHz, 0, output24kHz.Length);

            return output24kHz;
        }

        /// <summary>
        /// Resamples PCM audio from 24KHz to 16KHz using decimation
        /// </summary>
        /// <param name="input24kHz">Input PCM audio data at 24KHz (16-bit samples)</param>
        /// <returns>Resampled PCM audio data at 16KHz (16-bit samples)</returns>
        public static byte[] Resample24kTo16k(byte[] input24kHz)
        {
            if (input24kHz == null || input24kHz.Length == 0)
                return input24kHz;

            // Convert byte array to 16-bit samples
            int inputSampleCount = input24kHz.Length / 2;
            short[] inputSamples = new short[inputSampleCount];
            Buffer.BlockCopy(input24kHz, 0, inputSamples, 0, input24kHz.Length);

            // Calculate output sample count (16/24 = 0.6667x samples)
            int outputSampleCount = (int)(inputSampleCount * (16000.0 / 24000.0));
            short[] outputSamples = new short[outputSampleCount];

            // Resample using linear interpolation
            double ratio = 24000.0 / 16000.0; // 1.5

            for (int i = 0; i < outputSampleCount; i++)
            {
                double srcIndex = i * ratio;
                int srcIndexInt = (int)srcIndex;
                double fraction = srcIndex - srcIndexInt;

                if (srcIndexInt >= inputSampleCount - 1)
                {
                    outputSamples[i] = inputSamples[inputSampleCount - 1];
                }
                else
                {
                    // Linear interpolation between two samples
                    short sample1 = inputSamples[srcIndexInt];
                    short sample2 = inputSamples[srcIndexInt + 1];
                    outputSamples[i] = (short)(sample1 + (sample2 - sample1) * fraction);
                }
            }

            // Convert back to byte array
            byte[] output16kHz = new byte[outputSamples.Length * 2];
            Buffer.BlockCopy(outputSamples, 0, output16kHz, 0, output16kHz.Length);

            return output16kHz;
        }
    }
}

File: src/azure-communication-services-integrate-voice-live-api/python/audio_resampler.py
"""
Audio resampling utilities for converting between different sample rates.
"""
import numpy as np
from scipy import signal
import logging

logger = logging.getLogger(__name__)


class AudioResampler:
    """Audio resampling utilities for Voice Live API integration."""
    
    @staticmethod
    def resample_24k_to_16k(audio_bytes: bytes) -> bytes:
        """
        Resample audio from 24kHz to 16kHz to match ACS requirements.
        Voice Live API outputs 24kHz audio, but ACS expects 16kHz.
        
        Args:
            audio_bytes: PCM audio data at 24kHz, 16-bit, mono
            
        Returns:
            Resampled PCM audio data at 16kHz, 16-bit, mono
        """
        try:
            # Convert bytes to numpy array (16-bit PCM)
            audio_array = np.frombuffer(audio_bytes, dtype=np.int16)
            
            if len(audio_array) == 0:
                return audio_bytes
            
            # Resample from 24kHz to 16kHz
            # This is a ratio of 2:3, so we need to downsample
            num_samples_output = int(len(audio_array) * 16000 / 24000)
            
            # Use scipy.signal.resample for high-quality resampling
            resampled_array = signal.resample(audio_array, num_samples_output)
            
            # Convert back to int16 and then to bytes
            resampled_int16 = np.round(resampled_array).astype(np.int16)
            resampled_bytes = resampled_int16.tobytes()
            
            return resampled_bytes
            
        except Exception as e:
            logger.error(f"Error resampling audio: {e}")
            # Return original audio if resampling fails
            return audio_bytes
    
    @staticmethod
    def resample_16k_to_24k(audio_bytes: bytes) -> bytes:
        """
        Resample audio from 16kHz to 24kHz for Voice Live API.
        ACS outputs 16kHz audio, but Voice Live API expects 24kHz.
        
        Args:
            audio_bytes: PCM audio data at 16kHz, 16-bit, mono
            
        Returns:
            Resampled PCM audio data at 24kHz, 16-bit, mono
        """
        try:
            # Convert bytes to numpy array (16-bit PCM)
            audio_array = np.frombuffer(audio_bytes, dtype=np.int16)
            
            if len(audio_array) == 0:
                return audio_bytes
            
            # Resample from 16kHz to 24kHz
            # This is a ratio of 3:2, so we need to upsample
            num_samples_output = int(len(audio_array) * 24000 / 16000)
            
            # Use scipy.signal.resample for high-quality resampling
            resampled_array = signal.resample(audio_array, num_samples_output)
            
            # Convert back to int16 and then to bytes
            resampled_int16 = np.round(resampled_array).astype(np.int16)
            resampled_bytes = resampled_int16.tobytes()
            
            return resampled_bytes
            
        except Exception as e:
            logger.error(f"Error resampling audio 16k→24k: {e}")
            # Return original audio if resampling fails
            return audio_bytes
    
    @staticmethod
    def is_silent_audio(audio_bytes: bytes, threshold: float = 0.01) -> bool:
        """
        Check if audio data is silent (below threshold).
        
        Args:
            audio_bytes: PCM audio data
            threshold: RMS threshold for silence detection (0.0 - 1.0)
            
        Returns:
            True if audio is considered silent
        """
        try:
            if len(audio_bytes) == 0:
                return True
                
            # Convert to numpy array
            audio_array = np.frombuffer(audio_bytes, dtype=np.int16)
            
            # Calculate RMS (Root Mean Square)
            rms = np.sqrt(np.mean(audio_array.astype(np.float32) ** 2))
            
            # Normalize RMS to 0-1 range (32767 is max value for int16)
            normalized_rms = rms / 32767.0
            
            return normalized_rms < threshold
            
        except Exception as e:
            logger.error(f"Error checking audio silence: {e}")
            return False
